[transformer] BT enablement on fairseq - pytorch change

frank-wei opened the pull request:
The fairseq diff is split into two parts.
The first diff (this one)
This diff is about creating a mask left align function to check the mask condition for nested tensor. It is necessary for torchscript deployment.

The second diff (D37082681)
Fork the inference path inside the forward function. If loaded the checkpoint file and perform the inference, we will deploy BT. Otherwise, fairseq take the position.


Reviewed By: mikekgfb

Differential Revision: D36057338

erichan1 reviewed (approved):
LGTM!

jbschlosser reviewed (commented):
Confused about this - I just see this PR adding a `_nested_tensor_from_mask_left_aligned` op. Some questions:
* How does this related to the PR summary? AFAICT just adding this op won't enable its use.
* How does this differ from `_nested_tensor_from_mask` - does it just avoid checks by making the assumption the mask is left-aligned? If so, couldn't we avoid quite a bit of duplication?

frank-wei commented:
> Confused about this - I just see this PR adding a `_nested_tensor_from_mask_left_aligned` op. Some questions:
> 
> * How does this related to the PR summary? AFAICT just adding this op won't enable its use.
The diff is optimized to split into two parts now. Part1 only includes this op change and associated with D36057338. Part2 only includes fairseq change (D37082681) but depend on Part1. The internal CI tests show the op runs well.

> * How does this differ from `_nested_tensor_from_mask` - does it just avoid checks by making the assumption the mask is left-aligned? If so, couldn't we avoid quite a bit of duplication?
It is the front part of "_nested_tensor_from_mask" implementation. But the purpose is to help check the left aligned in advance.

Commit:
[transformer] BT enablement on fairseq - pytorch change

Summary:
The fairseq diff is split into two parts.
The first diff (this one)
This diff is about creating a mask left align function to check the mask condition for nested tensor. It is necessary for torchscript deployment.

The second diff (D37082681)
Fork the inference path inside the forward function. If loaded the checkpoint file and perform the inference, we will deploy BT. Otherwise, fairseq take the position.

Perf in summary:
Accuracy: accuracy loss due to the fp16, the maximum diff is around 0.009. If we set it to fp32, there is no accuracy loss
Perf: the current fairseq has similar speed as vanilla version. After the enablement, the speedup is similar to standalone BT test.
With batch size=64
For V100, the speedup reaches to 1.23x
For A100, the speedup reaches to 1.38x

After enable nested tensor,
For V100, the speedup reaches to 2.46x

Test Plan: In D37082681

Reviewed By: mikekgfb

Differential Revision: D36057338

fbshipit-source-id: 2b824481481b8972d168f2751afa77cc9b3cbe02
